File: cmd/benchmark/benchmark.go
package main

import (
	"bytes"
	"flag"
	"fmt"
	"io/ioutil"
	"log"
	"math"
	"os"
	"os/exec"
	"path/filepath"
	"regexp"
	"sort"
	"strconv"
	"strings"
	"time"
)

var startBranch string

func getCommit(version string) (string, error) {
	cmd := exec.Command("git", "rev-parse", "--short", version)
	out, err := cmd.Output()
	return strings.TrimSpace(string(out)), err
}

func checkout(version string) error {
	cmd := exec.Command("git", "checkout", version)
	cmd.Stdout = os.Stdout
	cmd.Stderr = os.Stderr
	return cmd.Run()
}

func describe(v string) string {
	cmd := exec.Command("git", "log", "-1", "--pretty=format:%s", v)
	out, err := cmd.Output()
	check(err)
	return strings.TrimSpace(string(out))
}

func workdirDiff() (string, error) {
	cmd := exec.Command("git", "diff")
	out, err := cmd.Output()
	return strings.TrimSpace(string(out)), err
}

func getBranch() (string, error) {
	cmd := exec.Command("git", "rev-parse", "--symbolic-full-name", "HEAD")
	out, err := cmd.Output()
	branch := strings.TrimSpace(string(out))
	branch = strings.TrimPrefix(branch, "refs/heads/")
	return branch, err
}

func getCPUType() (string, error) {
	content, err := ioutil.ReadFile("/proc/cpuinfo")
	if err != nil {
		return "", err
	}

	for _, l := range strings.Split(string(content), "\n") {
		fields := strings.SplitN(l, "\t: ", 2)
		if fields[0] == "model name" {
			return fields[1], nil
		}
	}

	return "unknown", nil
}

func getCPUSpeedsKhz() (speeds map[string]int, err error) {
	keys := []string{"scaling_max_freq",
		"scaling_min_freq",
		"cpuinfo_min_freq",
		"cpuinfo_max_freq",
	}

	speeds = make(map[string]int)
	for _, k := range keys {
		content, err := ioutil.ReadFile("/sys/devices/system/cpu/cpu0/cpufreq/" + k)
		if err != nil {
			return nil, err
		}
		spd, err := strconv.Atoi(strings.TrimSpace(string(content)))
		if err != nil {
			return nil, err
		}
		speeds[k] = spd
	}
	return speeds, nil
}

func setCPUSpeed(minKhz, maxKhz int) error {
	min := fmt.Sprintf("%dkhz", minKhz)
	max := fmt.Sprintf("%dkhz", maxKhz)
	cmd := exec.Command("sudo", "-n", "cpupower", "frequency-set", "-u", max, "-d", min)
	cmd.Stdout = os.Stdout
	cmd.Stderr = os.Stderr
	log.Printf("running %s", cmd.Args)
	if err := cmd.Run(); err != nil {
		return err
	}
	return nil
}

func Error(msg string) {
	fmt.Println(msg)
	if startBranch != "" {
		fmt.Println("back to branch", startBranch)
		checkout(startBranch)
	}
	os.Exit(1)
}

func check(err error) {
	if err != nil {
		Error(err.Error())
	}
}

func sanityCheck() {
	diff, err := workdirDiff()
	check(err)
	if diff != "" {
		Error("should be clean")
	}
	content, err := ioutil.ReadFile("config.make")
	check(err)
	if !bytes.Contains(content, []byte("-O2")) {
		Error("need -O2 for benchmarking")
	}
}

func Shell(shcmd string) error {
	cmd := exec.Command("/bin/sh", "-c", shcmd)
	cmd.Stdout = os.Stdout
	cmd.Stderr = os.Stderr
	return cmd.Run()
}

func buildVersion(v string) error {
	if err := checkout(v); err != nil {
		return err
	}
	if content, err := ioutil.ReadFile("config.make"); err != nil {
		return err
	} else if !strings.Contains(string(content), "-O2") {
		return fmt.Errorf("config.mark is missing -O2")
	}
	bin := fmt.Sprintf("binaries/lilypond.%s", v)
	callBin := fmt.Sprintf("out/bin/lilypond.%s", v)
	if _, err := os.Lstat(bin); err != nil {
		if err := Shell(fmt.Sprintf("make -C flower/ clean && make -C lily/ clean && make -C flower -j4 && make -C lily/ -j4 && cp lily/out/lilypond %s", bin)); err != nil {
			return err
		}
	}
	os.Remove(callBin)
	if err := os.Link(bin, callBin); err != nil {
		return err
	}

	return nil
}

type runResults struct {
	mem  float64
	time float64
}

type allResults struct {
	baseMem  []float64
	versMem  []float64
	baseTime []float64
	versTime []float64
	runCount int
	args     []string
}

var (
	timeRe = regexp.MustCompile(`User time \(seconds\): ([0-9.]+)`)
	memRe  = regexp.MustCompile(`Maximum resident set size.*: ([0-9.]+)`)
)

func commandId(args []string) string {
	return regexp.MustCompile("[^a-zA-Z0-9-_]").ReplaceAllString(time.Now().Format(time.RFC3339)+strings.Join(args, ""), "")
}

func benchmark(v1, v2 string, count int, outDir string, args []string) allResults {
	mem := map[string][]float64{}
	time := map[string][]float64{}

	for i := 0; i < count; i++ {
		for _, v := range []string{v1, v2} {
			out := fmt.Sprintf("%s/%s-%s.%d.txt", outDir, commandId(args), v, i)

			check(checkout(v))
			cmd := fmt.Sprintf("/usr/bin/time -v out/bin/lilypond.%s %s >& %s", v, strings.Join(args, " "), out)
			log.Println("running", cmd)
			check(Shell(cmd))

			content, err := ioutil.ReadFile(out)
			check(err)
			cstr := string(content)
			match := timeRe.FindStringSubmatch(cstr)
			t, err := strconv.ParseFloat(match[1], 64)
			check(err)
			match = memRe.FindStringSubmatch(cstr)
			m, err := strconv.ParseFloat(match[1], 64)
			check(err)

			mem[v] = append(mem[v], m)
			time[v] = append(time[v], t)
		}
	}

	return allResults{
		baseMem:  mem[v1],
		baseTime: time[v1],
		versMem:  mem[v2],
		versTime: time[v2],
		args:     args,
		runCount: count,
	}
}

func stddev(fs []float64) float64 {
	a := mean(fs)
	tot := 0.0
	for _, f := range fs {
		tot += (f - a) * (f - a)
	}
	return math.Sqrt(tot / float64(len(fs)-1))
}

func mean(fs []float64) float64 {
	tot := 0.0
	for _, f := range fs {
		tot += f
	}
	return tot / float64(len(fs))
}

func analyzeData(base []float64, data []float64, name string, version string) string {
	sort.Float64s(base)
	sort.Float64s(data)

	baseMed := base[len(base)/2]
	versMed := data[len(data)/2]

	delta := versMed - baseMed
	baseStddev := stddev(base)
	dataStddev := stddev(data)

	qual := "neutral"
	if math.Abs(delta) > baseStddev+dataStddev {
		if name == "mem" {
			if delta > 0 {
				qual = "fatter"
			} else {
				qual = "leaner"
			}
		}
		if name == "time" {
			if delta > 0 {
				qual = "slower"
			} else {
				qual = "faster"
			}
		}
	}

	return fmt.Sprintf(`    %s delta: %f (stddev %f %f n=%d)
    %s delta: %f %% (%s is %s)`, name, delta, baseStddev, dataStddev, len(base), name, 100.0*delta/baseMed, version, qual)
}

func analyze(base, vers string, res allResults) string {
	speeds, _ := getCPUSpeedsKhz()
	cpu, _ := getCPUType()
	if strings.Contains(cpu, "@") {
		fs := strings.SplitN(cpu, "@", 2)
		cpu = fs[0]
	}

	r := fmt.Sprintf(`benchmark for arguments: %s

%s at %d Mhz

raw data (%s):
   %v %v
raw data (%s):
   %v %v
`, res.args,
		cpu, speeds["scaling_min_freq"]/1000,
		base, res.baseMem, res.baseTime,
		vers, res.versMem, res.versTime)

	r += fmt.Sprintf(`%s - %s
    baseline %s - %s
    args %s
%s
%s
`, vers, describe(vers), base, describe(base), res.args,
		analyzeData(res.baseMem, res.versMem, "mem", vers),
		analyzeData(res.baseTime, res.versTime, "time", vers))

	return r
}

func main() {
	version := flag.String("version", "HEAD", "version to test")
	baseline := flag.String("baseline", "", "baseline")
	runCount := flag.Int("run_count", 3, "run count")
	outDir := flag.String("out", "benchmark-results", "out dir")

	flag.Parse()
	speeds, err := getCPUSpeedsKhz()
	check(err)

	if min, max := speeds["scaling_min_freq"], speeds["scaling_max_freq"]; min != max {
		if err := setCPUSpeed(speeds["cpuinfo_max_freq"]/2, speeds["cpuinfo_max_freq"]/2); err != nil {
			log.Fatalf("setCPUSpeed %v", err)
		}
		defer setCPUSpeed(speeds["cpuinfo_min_freq"], speeds["cpuinfo_max_freq"])
	} else {
		log.Printf("CPU freq %d Mhz", min/1000)
	}
	if *baseline == "" {
		*baseline, err = getCommit(*version + "^")
		check(err)
	}
	*baseline, err = getCommit(*baseline)
	check(err)
	*version, err = getCommit(*version)
	check(err)
	startBranch, err = getBranch()
	check(err)
	log.Println("startbranch is", startBranch)
	sanityCheck()

	for _, v := range []string{*baseline, *version} {
		check(buildVersion(v))
	}

	result := benchmark(*baseline, *version, *runCount, *outDir, flag.Args())
	log.Println(result)
	summary := analyze(*baseline, *version, result)
	log.Println(summary)
	check(ioutil.WriteFile(filepath.Join(*outDir, fmt.Sprintf("%s-v%s-base%s", commandId(flag.Args()),
		*version, *baseline)), []byte(summary), 0644))
	if startBranch != "" {
		checkout(startBranch)
	}
}

File: benchmark.py
#!/usr/bin/python


"""
Script for structured benchmarking of lilypond binaries. To use, compile like this:


$ make -C lily && test -z "$(git diff )" && cp lily/out/lilypond binaries/lilypond.$(git rev-parse  --short HEAD) && ln binaries/* out/bin/

then

$ python scripts/auxiliar/benchmark.py -n3  -v HEAD^ -v HEAD -- input/regression/mozart-hrn-3.ly


The first -v should be the baseline, the second should be what you want to test.
"""

import math
import re
import os
import sys
import statistics
import getopt

def error(msg):
  sys.stdout.write("%s\n" % msg)
  sys.exit(1)

def system(cmd, allow_error=False):
  print("running %s" % cmd)
  status = os.system(cmd)
  if status and not allow_error:
    error("failure running %s" % cmd)
  return status

class Options:
  pass


def parse_cmdline():
  options = Options()
  options.run_count = 3
  options.versions = []
  options.descriptions = {}
  options.out_dir = "benchmark-results"

  opts, args = getopt.getopt(sys.argv[1:], "n:b:v:",[])

  baseline = ''
  testversion = ''
  for (o, a) in opts:
    if o == '-n':
      options.run_count = int(a)
    if o == '-b':
      baseline = a
    if o == '-v':
      commit = os.popen("git rev-parse --short %s" % a).read().strip()
      testversion = commit

  if not baseline:
    baseline = os.popen("git rev-parse --short %s^" % testversion).read().strip()

  options.versions = [baseline, testversion]
  for v in options.versions:
      descr = os.popen("git log -1 --pretty=format:%%s %s" % v).read()
      options.descriptions[v] = descr

  options.args = args
  return options

def sanity_check():
  if '-O2' not in open('config.make').read():
    error("need -O2 in configuration")

  diff = os.popen("git diff").read()
  if diff:
    error("must have clean tree")

def get_branch():
  cur_branch = ""
  full_ref = os.popen("git rev-parse --symbolic-full-name HEAD").read()
  if full_ref.startswith("refs/heads/"):
    cur_branch = full_ref[len("refs/heads/"):]
  return cur_branch

def build_versions(versions):
  for v in versions:
    if not os.path.exists("binaries/lilypond.%s" % v):
      system("git checkout %s" % v)
      system("make -C flower/ clean && make -C lily/ clean && make -C flower && make -C lily/ -j4")
      system("cp lily/out/lilypond binaries/lilypond.%s" % v)
    if not os.path.exists("out/bin/lilypond.%s" % v):
      system("ln binaries/lilypond.%s out/bin/lilypond.%s" % (v, v))


def command_id (options):
  return  re.sub('[^a-zA-Z0-9_-]', '', '-'.join(options.args))

def run_timings(options):
  os.makedirs(options.out_dir, exist_ok=True)

  memory_results = {}
  timing_results = {}
  for v in options.versions:
    timing_results[v] = []
    memory_results[v] = []


  for run in range(0, options.run_count):
    for v in options.versions:
      out = '%s/%s-%s.%d.txt' % (options.out_dir, command_id(options), v, run)
      system("git checkout %s" % v)
      status  = system("/usr/bin/time -v out/bin/lilypond.%s %s >& %s" % (v, ' '.join(options.args), out))

      if status:
        system("git checkout %s" % cur_branch)
        error("lilypond run failed")

      for l in open(out).readlines():
        m = re.search('User time \(seconds\): ([0-9.]+)', l)
        if m:
          timing_results[v].append(float(m.group(1)))

        m = re.search('Maximum resident set size.*: ([0-9.]+)', l)
        if m:
          memory_results[v].append(float(m.group(1)))
  return memory_results, timing_results

def stddev(l):
  if len(l) > 1:
    return statistics.stdev(l)
  return float("+inf")

def abs_summary(options, memory_results, timing_results):
  result = ""

  result += "benchmark for arguments: %s\n" % ' '.join(options.args)

  result += "raw data:\n  mem %s\n  time %s\n" % (memory_results, timing_results)

  for v in options.versions:
    timings = timing_results[v]
    memory = memory_results[v]
    result +=("Version %s: %s\n" % (v, options.descriptions[v]))
    result += ("  time avg %f\n" % statistics.mean(timings))
    result += ("  time med %f\n" % statistics.median(timings))
    result += ("  time stddev %.3f\n" % stddev(timings))

    result += ("  mem  avg %f\n" % statistics.mean(memory))
    result += ("  mem  med %f\n" % statistics.median(memory))
    result += ("  mem  stddev %.3f\n" % stddev(memory))
  return result

def delta_summary_metric(base_version, version, run_count, metric_name, metric):
  base = metric[base_version]
  data = metric[version]
  result = ""
  delta =  statistics.median(data) - statistics.median(base)
  if metric_name == "memory":
    prec = 0
    qual = ((delta > 0) and "fatter") or "leaner"
  elif metric_name == "time":
    prec = 2
    qual = ((delta > 0) and "slower") or "faster"
  else:
    assert 0, metric_name

  if math.fabs(delta) < (stddev(base) + stddev(base)):
    qual = "neutral"

  result += "  %s: med diff %.*f (stddevs %.*f %.*f, n=%d)\n" % (metric_name, prec, delta, prec, stddev(base), prec, stddev(data), run_count)
  result += "  %s: med diff %.1f %% (%s is %s)\n" % (metric_name, delta / statistics.median(base) * 100.0, version, qual)
  return result

def delta_summary_version(options, v, memory_results, time_results):
  result = ""
  result += "%s - %s\n" %(v, options.descriptions[v])
  result += "  baseline: %s %s\n" %(options.versions[0], options.descriptions[options.versions[0]])
  result += "  args: %s\n" %(' '.join(options.args))
  result += delta_summary_metric(options.versions[0], v, options.run_count, "memory", memory_results)
  result += delta_summary_metric(options.versions[0], v, options.run_count, "time", time_results)
  return result



def main():
  options = parse_cmdline()
  print(options.descriptions)
  sanity_check()

  build_versions(options.versions)

  mem, time = run_timings(options)

  result = abs_summary(options, mem, time)
  for v in options.versions[1:]:
    result += delta_summary_version(options, v, mem, time)

  open("%s/%s-%s-summary.txt" % (options.out_dir, command_id(options), '-'.join(options.versions[1:])), 'w').write(result)

  sys.stdout.write(result)
  print()

cur_branch = get_branch()
try:
  main()
finally:
  system("git checkout %s" % cur_branch)
